### PR TITLE
Add FlashAttention2 for XLM-RoBERTa

### PR DESCRIPTION
**Pull request by DavidAfonsoValente**
# What does this PR do?

Fixes #27957

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by amyeroberts** [comment]
@DavidAfonsoValente Thanks for opening a PR and adding FA2 for this model! Let us know when the PR is ready to be reviewed 🤗

**Comment by DavidAfonsoValente** [comment]
> @DavidAfonsoValente Thanks for opening a PR and adding FA2 for this model! Let us know when the PR is ready to be reviewed 🤗

Thanks! I believe it's ready to be reviewed.

**Comment by DavidAfonsoValente** [comment]
The tests that are failing are due to the changes made in the functions that are labeled as:
# Copied from transformers.models.roberta.modeling_roberta.RobertaSelfAttention with Roberta->XLMRoberta

What is the standard procedure when I have made changes to these functions in order to accommodate the new feature?

**Comment by amyeroberts** [comment]
@DavidAfonsoValente If you click on the CI runs, you'll see the error messages relating to the code quality failures. These detail how to resolve the issue: run `make fix-copies` and push the changes. 

If another model is copying this model's attention classes, then it's great because you're adding FA2 for two models 🥳  You just need to apply the equivalent changes to the model that's copying.

File: tests/models/xlm_roberta/test_modeling_xlm_roberta.py [hunk 1]
```diff
@@ -67,3 +89,350 @@ def test_xlm_roberta_large(self):
         self.assertEqual(output.shape, expected_output_shape)
         # compare the actual values for a slice of last dim
         self.assertTrue(torch.allclose(output[:, :, -1], expected_output_values_last_dim, atol=1e-3))
+
+
+class XLMRobertaModelTester(object):
```
Comment (amyeroberts) [inline tests/models/xlm_roberta/test_modeling_xlm_roberta.py:94]: nit

```suggestion
class XLMRobertaModelTester:
```

File: tests/models/xlm_roberta/test_modeling_xlm_roberta.py [hunk 2]
```diff
@@ -67,3 +89,350 @@ def test_xlm_roberta_large(self):
...
+        self.parent.assertEqual(result.logits.shape, (self.batch_size, self.num_choices))
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (config, input_ids, input_mask, sequence_labels, token_labels, choice_labels) = config_and_inputs
+        inputs_dict = {"input_ids": input_ids, "attention_mask": input_mask}
+        return config, inputs_dict
+
+
+class XLMRobertaModelTest(ModelTesterMixin, PipelineTesterMixin, unittest.TestCase):
```
Comment (amyeroberts) [inline tests/models/xlm_roberta/test_modeling_xlm_roberta.py:269]: Thanks for adding this!

File: docs/source/en/model_doc/xlm-roberta.md [hunk 3]
```diff
@@ -113,6 +113,35 @@ A list of official Hugging Face and community (indicated by 🌎) resources to h
...
+>>> model = AutoModel.from_pretrained("XLM-RoBERTa-base", torch_dtype=torch.float16, attn_implementation="flash_attention_2")
+
+>>> text = "Replace me by any text you'd like."
+
+>>> encoded_input = tokenizer(text, return_tensors='pt').to(device)
+>>> model.to(device)
+
+>>> output = model(**encoded_input)
+```
+
```
Comment (amyeroberts) [inline docs/source/en/model_doc/xlm-roberta.md:144]: We should performance numbers here too e.g. like here for gpt neo: https://github.com/huggingface/transformers/blob/abbffc4525566a48a9733639797c812301218b83/docs/source/en/model_doc/gpt_neo.md#expected-speedups

**Review by amyeroberts — COMMENTED** [review]
Awesome work - thanks for adding!

Just a few tiny comments and then we're good to merge 🤗

File: docs/source/en/model_doc/xlm-roberta.md (hunk 3 above)
Comment (DavidAfonsoValente) [reply docs/source/en/model_doc/xlm-roberta.md:144]: I can't get flash_attn running on my setup, is there another way to run these benchmarks?

Comment (amyeroberts) [reply docs/source/en/model_doc/xlm-roberta.md:144]: What is the issue you're encountering when setting up FA2? We can run benchmarks if it's needed in terms of machine size, etc., but you'll need to be able to run FA2 on your machine to run the tests and verify this addition is correct.

**Comment by hackyon** [comment]
Hello there! 

I'm working on integrating scaled_dot_product_attention to BERT #28802, and there might be some merge conflicts with this change. 

Most of the changes I'm making will propagate through to XML-RoBERTa (through fix-copies). From what I can, our approaches are a bit different from this in 2 areas:
1. I've kept the `#Copies from` line at the class level, so any changes to BERT's init() will be propagated through to the fix-copies classes. I've modified the `#Copies from` line to use the corresponding ATTENTION_CLASS as necessary.
2. I've created a BertSdpaSelfAttention to go with BertSelfAttention, instead of creating a BertSdpaAttention. The rationale behind this is that the code in BertSdpaSelfAttention is adapted from the code in BertSelfAttention, so I figured it's better to name it as such.

Let me know if you have any questions about these. I think we should discuss about which approach is better and adopt it. Would you mind looking through #28802 and let me know what you think?

**Comment by DavidAfonsoValente** [comment]
> Hello there!
> 
> I'm working on integrating scaled_dot_product_attention to BERT #28802, and there might be some merge conflicts with this change.
> 
> Most of the changes I'm making will propagate through to XML-RoBERTa (through fix-copies). From what I can, our approaches are a bit different from this in 2 areas:
> 
> 1. I've kept the `#Copies from` line at the class level, so any changes to BERT's init() will be propagated through to the fix-copies classes. I've modified the `#Copies from` line to use the corresponding ATTENTION_CLASS as necessary.
> 2. I've created a BertSdpaSelfAttention to go with BertSelfAttention, instead of creating a BertSdpaAttention. The rationale behind this is that the code in BertSdpaSelfAttention is adapted from the code in BertSelfAttention, so I figured it's better to name it as such.
> 
> Let me know if you have any questions about these. I think we should discuss about which approach is better and adopt it. Would you mind looking through #28802 and let me know what you think?

It seems like your implementation doesnt create complicated conflitcts, it should be simple to merge, what do you think?

**Comment by hackyon** [comment]
It should be simple to merge. I think the main question is whether or not we want to use the XLM_ROBERTA_SELF_ATTENTION_CLASSES or XLM_ROBERTA_ATTENTION_CLASSES approach?

**Comment by DavidAfonsoValente** [comment]
I think in order to be consistent with other models we should keep the name XLM_ROBERTA_ATTENTION_CLASSES since this is how other models have their attention classes named.

**Comment by hackyon** [comment]
Thanks for your input.

I am leaning towards using the "SelfAttention" convention in this case, because XLMRobertaFlashAttention2 actually tries to copy the logic of both XLMRobertaAttention and XLMRobertaSelfAttention into one. I think it's cleaner to have a XLMRobertaSelfFlashAttention2 that mirrors XLMRobertaSelfAttention, and then reuse the logic inside XLMRobertaAttention for both types of self attentions.

Reusing the code in XLMRobertaAttention should help avoid some bugs. Judging from the existing code, I think there's already a bit of an issue with the way self.output() is called in XLMRobertaFlashAttention2 (I think you need more than the call to self.output.dense(), otherwise you're missing the dropout and LayerNorm).

**Comment by DavidAfonsoValente** [comment]
I believe the layernorm is executed after the attention layer in the transformer block so that shouldn´t be a problem:
`self.attention = DISTILBERT_ATTENTION_CLASSES[config._attn_implementation](config)
 self.sa_layer_norm = nn.LayerNorm(normalized_shape=config.dim, eps=1e-12)`
 As for the dropout, could you please clarify? I assume you're referring to the dropout that is called already in:
`flash_attn_func(
                query_states, key_states, value_states, dropout, softmax_scale=softmax_scale, causal=causal
            )`

**Comment by DavidAfonsoValente** [comment]
Okk, maybe it'll be easier if you merge your PR and I work on top of your version.

**Comment by michaelshekasta** [comment]
@amyeroberts @DavidAfonsoValente @hackyon Why hasn't this PR been merged yet? How can I help with that?

**Comment by hackyon** [comment]
@michaelshekasta

I have this #30510 open to add scaled dot product attention (SDPA) to RoBERTa-based models, and is still under review. When available, the SDPA could run on a FA2 backend. You could very well try to add FA2 to it as well.

**Comment by michaelshekasta** [comment]
> @michaelshekasta
> 
> I have this #30510 open to add scaled dot product attention (SDPA) to RoBERTa-based models, and is still under review. When available, the SDPA could run on a FA2 backend. You could very well try to add FA2 to it as well.

@hackyon Thank you for your feedback! As far as I'm aware, SDPA isn't Flash Attention 2, and it's currently in beta, correct? So, will HF support it? Or am I missing something?

I'd be happy to contribute, but I'm not sure how. Could we discuss this on the HF Discord server, perhaps in the NLP channel?

**Comment by hackyon** [comment]
Newer versions of SDPA actually has FA2 integration: https://pytorch.org/blog/pytorch2-2. With that said, you could probably add a FA2 option as well if you'd like (some of the models have that as well), but I'm not HG maintainer so take my opinion with a grain of salt.

SDPA is already supported in a variety of HG models.

**Comment by michaelshekasta** [comment]
Thank you once more for your input, @hackyon!

I'm interested in implementing XLM-RoBERTa with FA2, and I believe integrating it into Hugging Face (HF) could be the way forward. From what I've gathered, this pull request appears to align with my goals.

While I understand you're not directly involved in HF maintenance, your prior work suggests you're knowledgeable in this area. Could you advise on the best course of action? Should I wait for your pull request, or would utilizing the flash_attn library be a viable alternative?

**Comment by michaelshekasta** [comment]
@hackyon thanks again! you are awesome!
I think that I don't really understand the issue with Copy.from..  can you explain me more or refer me to resource?

**Comment by michaelshekasta** [comment]
@hackyon, upon revisiting the thread, I've come to the conclusion that transitioning from SDPA to XLMR seems to be the most appropriate course of action. Given that PyTorch has already implemented FA2 with SDPA, I don't see the necessity of replicating it for XLMR. Is there anything I can assist with to facilitate the seamless integration of this pull request?

**Comment by michaelshekasta** [comment]
@hackyon Let's talk in the email shkasta at post dot bgu dot ac dot il

**Comment by aikangjun** [comment]
hi， I want to know whether this job has been merged into 4.42.2?

**Comment by amyeroberts** [comment]
@aikangjun This PR wasn't merged - it closed because of inactivity
